### PR TITLE
Added assert for opengles thread safety

### PR DESCRIPTION
**Pull request by gaaclarke**
This assert would have blocked the following bug: https://github.com/flutter/flutter/issues/158535

This shouldn't be landed until https://github.com/flutter/engine/pull/56573 lands.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/renderer/backend/gles/proc_table_gles.h [hunk 1]
```diff
@@ -99,6 +101,14 @@ struct GLProc {
   ///
   bool log_calls = false;
 
+  //----------------------------------------------------------------------------
+  /// Whether the OpenGL call asserts it is only used from / one thread in
+  /// IMPELLER_DEBUG builds.
```
Comment (jonahwilliams) [inline impeller/renderer/backend/gles/proc_table_gles.h:106]: IMPELLER_DEBUG is debug/profile mode.  !defined(NDEBUG) means unopt local engine IIRC, that should be fine.

**Comment by gaaclarke** [comment]
> I don't think this is correct. We can call blitFramebuffer from the raster thread as part of rendering and still separately use blitFramebuffer from the iio thread for image resize.

The only symbol that is set up to enforce this is `glUseProgram`.  Under no circumstances today is it valid to use `glUseProgram` from any thread but the raster thread.  No meaningful drawing can be done without `glUseProgram`, blitting and uploading texture data is fine.

**Comment by jonahwilliams** [comment]
oh duh, kk

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by gaaclarke** [comment]
It might break benchmarks on device lab though, so that's good.

**Comment by jonahwilliams** [comment]
It won't because those don't run unopt builds.

**Comment by gaaclarke** [comment]
> It won't because those don't run unopt builds.

Hmm, that sounds like we should potentially move this to just `#if IMPELLER_DEBUG` then, no?

**Comment by jonahwilliams** [comment]
We should have expensive debug checks active. We've already found that adding too much overhead via validation layers negatively impacts developer experience.

**Comment by jonahwilliams** [comment]
should = shouldn't 🤦

**Pull request merged**